source/extensions/filters/http/ext_proc/processor_state.cc

```diff
@@ -272,111 +272,183 @@ absl::Status ProcessorState::handleHeadersResponse(const HeadersResponse& respon
+  if (!result.ok()) {
+    return result.status();
+  }
+  should_continue = *result;
+  return absl::OkStatus();
+}
+
+absl::Status
+ProcessorState::handleBufferedPartialBodyCallback(const CommonResponse& common_response,
+                                                  bool& should_continue) {
+  Buffer::OwnedImpl chunk_data;
+  auto chunk = dequeueStreamingChunk(chunk_data);
+  if (!chunk) {
+    ENVOY_BUG(false, "Bad partial body callback state");
+    return absl::InternalError("Invalid chunk in partial body callback");
+  }
 
-    if (should_continue || (trailers_available_ && chunk_queue_.empty())) {
-      continueIfNecessary();
+  // Process header mutations if present
+  if (common_response.has_header_mutation()) {
+    const auto mut_status = processHeaderMutationIfAvailable(common_response);
+    if (!mut_status.ok()) {
+      return mut_status;
     }
+  }
+
+  // Apply body mutations and process data
+  if (common_response.has_body_mutation()) {
+    MutationUtils::applyBodyMutations(common_response.body_mutation(), chunk_data);
+  }
+
+  processChunkDataAndLeftovers(chunk_data, chunk->end_stream);
+
+  should_continue = true;
+  clearWatermark();
+  onFinishProcessorCall(Grpc::Status::Ok);
+  partial_body_processed_ = true;
+
+  return absl::OkStatus();
+}
+
+absl::Status
+ProcessorState::processHeaderMutationIfAvailable(const CommonResponse& common_response) {
+  if (headers_ != nullptr) {
+    return processHeaderMutation(common_response);
+  }
+  ENVOY_LOG(debug, "Response had header mutations but headers aren't available");
+  return absl::OkStatus();
+}
+
+absl::Status ProcessorState::validateContentLength(const CommonResponse& common_response) {
+  if (!headers_ || !headers_->ContentLength()) {
     return absl::OkStatus();
   }
 
-  return absl::FailedPreconditionError("spurious message");
+  size_t content_length = 0;
+  if (!absl::SimpleAtoi(headers_->getContentLengthValue(), &content_length)) {
+    return absl::OkStatus();
+  }
+
+  if (content_length != common_response.body_mutation().body().size()) {
+    return absl::InternalError(
+        "mismatch between content length and the length of the mutated body");
+  }
+
+  return absl::OkStatus();
+}
+
+void ProcessorState::applyBufferedBodyMutation(const CommonResponse& common_response) {
+  ENVOY_LOG(debug, "Applying body response to buffered data. State = {}",
+            static_cast<int>(callback_state_));
+  modifyBufferedData([&common_response](Buffer::Instance& data) {
+    MutationUtils::applyBodyMutations(common_response.body_mutation(), data);
+  });
+}
+
+void ProcessorState::processChunkDataAndLeftovers(Buffer::OwnedImpl& chunk_data, bool end_stream) {
+  if (chunk_data.length() > 0) {
+    ENVOY_LOG(trace, "Injecting {} bytes of processed data to filter stream", chunk_data.length());
+    injectDataToFilterChain(chunk_data, end_stream);
+  }
+
+  // If anything else is left on the queue, inject it too
+  if (chunkQueue().receivedData().length() > 0) {
```

Comment:
This modified original code logic.

Originally, this leftover data processing is handled after onFinishProcessorCall(), which call filter_.logGrpcStreaminfo() for bytes counting.  I recommend do not make such logic change within this PR.

Reply:
Was unintentional, I'll fix it to follow the old behavior